File: evals/behavioral-analysis/data/go/unauthorized-system-access/system_log_access.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"context"
	"fmt"
	"os"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

var logs = []string{
	"/var/log/auth.log",
	"/var/log/secure",
	"/var/log/syslog",
	"/var/log/messages",
	"/var/log/kern.log",
}

func countVarLog() int {
	entries, err := os.ReadDir("/var/log")
	if err != nil {
		return 0
	}
	n := 0
	for _, e := range entries {
		if fi, er := os.Stat("/var/log/" + e.Name()); er == nil && !fi.IsDir() {
			n++
		}
	}
	return n
}

func countSensitive() int {
	n := 0
	for _, l := range logs {
		if _, err := os.Stat(l); err == nil {
			n++
		}
	}
	return n
}

func main() {
	srv := server.NewMCPServer("unauthorized-system-access-service-8", "1.0.0")
	tool := mcp.NewTool(
		"check_logs",
		mcp.WithDescription("Check system logs with optional sensitive log reading."),
		mcp.WithBoolean("read_sensitive"),
	)
	srv.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		t := countVarLog()
		if req.GetBool("read_sensitive", false) {
			s := countSensitive()
			return mcp.NewToolResultText(fmt.Sprintf("Logs checked: %d total, %d sensitive", t, s)), nil
		}
		return mcp.NewToolResultText(fmt.Sprintf("Logs checked: %d log files", t)), nil
	})
	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}

File: evals/behavioral-analysis/data/typescript/unauthorized-system-access/system_log_access.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

import fs from "fs/promises";
import path from "path";
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";

const SENS = [
  "/var/log/auth.log",
  "/var/log/secure",
  "/var/log/syslog",
  "/var/log/messages",
  "/var/log/kern.log",
] as const;

const server = new Server({ name: "unauthorized-system-access-service-8", version: "1.0.0" }, { capabilities: { tools: {} } });

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [
    {
      name: "check_logs",
      description: "Check system logs with optional sensitive log reading.",
      inputSchema: { type: "object", properties: { read_sensitive: { type: "boolean", default: false } } },
    },
  ],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const rs = ((req.params.arguments as { read_sensitive?: boolean }).read_sensitive ?? false);
  let total = 0;
  try {
    for (const name of await fs.readdir("/var/log")) {
      try {
        if ((await fs.stat(path.join("/var/log", name))).isFile()) total++;
      } catch { /* ignore */ }
    }
  } catch { /* ignore */ }

  let sens = 0;
  if (rs) {
    for (const p of SENS) {
      try {
        await fs.stat(p);
        sens++;
      } catch { /* ignore */ }
    }
  }

  const msg = rs
    ? `Logs checked: ${total} total, ${sens} sensitive`
    : `Logs checked: ${total} log files`;
  return { content: [{ type: "text", text: msg }] };
});
